### ISSUE TITLE
Conflicting nullability specifier warnings

### ISSUE THREAD

**Issue by mkujalowicz** [openid/AppAuth-iOS#60]
There are couple of warnings when bridging the library with Swift 3 (Xcode 8)

```
AppAuth-iOS/Source/OIDAuthState.h:144:1: warning: conflicting nullability specifier on return types, 'nullable' conflicts with existing specifier 'nonnull'
- (nullable instancetype)init NS_UNAVAILABLE;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.1.sdk/usr/include/objc/NSObject.h:60:1: note: previous declaration is here
- (instancetype)init
^

AppAuth-iOS/Source/OIDAuthorizationResponse.h:106:1: warning: conflicting nullability specifier on return types, 'nullable' conflicts with existing specifier 'nonnull'
- (nullable instancetype)init NS_UNAVAILABLE;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.1.sdk/usr/include/objc/NSObject.h:60:1: note: previous declaration is here
- (instancetype)init
^

AppAuth-iOS/Source/OIDAuthorizationService.h:84:1: warning: conflicting nullability specifier on return types, 'nullable' conflicts with existing specifier 'nonnull'
- (nullable instancetype)init NS_UNAVAILABLE;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.1.sdk/usr/include/objc/NSObject.h:60:1: note: previous declaration is here
- (instancetype)init
^

AppAuth-iOS/Source/OIDServiceDiscovery.h:358:1: warning: conflicting nullability specifier on return types, 'nullable' conflicts with existing specifier 'nonnull'
- (nullable instancetype)init NS_UNAVAILABLE;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.1.sdk/usr/include/objc/NSObject.h:60:1: note: previous declaration is here
- (instancetype)init
^

AppAuth-iOS/Source/OIDTokenRequest.h:114:1: warning: conflicting nullability specifier on return types, 'nullable' conflicts with existing specifier 'nonnull'
- (nullable instancetype)init NS_UNAVAILABLE;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.1.sdk/usr/include/objc/NSObject.h:60:1: note: previous declaration is here
- (instancetype)init
^

AppAuth-iOS/Source/OIDTokenResponse.h:97:1: warning: conflicting nullability specifier on return types, 'nullable' conflicts with existing specifier 'nonnull'
- (nullable instancetype)init NS_UNAVAILABLE;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.1.sdk/usr/include/objc/NSObject.h:60:1: note: previous declaration is here
- (instancetype)init
^
```

Some other initializers also have questionable nullability specifiers, e.g.
OIDTokenRequest:
```
- (nullable instancetype)initWithConfiguration:(OIDServiceConfiguration *)configuration …
```

This initializer should be marked as nullable only when there are some circumstances where the initializer can return nil, which is not the case here. This initializer always returns value from NSObject’s init, which never returns nil:
https://developer.apple.com/reference/objectivec/nsobject/1418641-init

With the current annotations, many initializers are marked as “failable” in Swift, e.g.:
let tokenRequest = OIDTokenRequest(…)
initializes token request as optional and requires unnecessary unwrapping

**Comment by jchnxu** [comment]
@mkujalowicz  Could you share your project please? I'm having some trouble bridging the static library in swift too.

**Comment by slizeray** [comment]
@jchnxu I am having the same issue, just use Cocoapods as documented in the readme, I think it is the easiest....

**Comment by jchnxu** [comment]
@slizeray I know and I got them working. It's just I'm trying to make a swift demo and send a pull request. And the project structure is different from the current pod development structure. So I might need to refactor the whole project which is super complicated and may be rejected, or attack this problem head on.

**Comment by mkujalowicz** [comment]
@jchnxu I have pushed a simple example how to integrate AppAuth with Swift, specifically swift unit tests:
https://github.com/mkujalowicz/AppAuth-iOS/tree/swift_tests

I've added additional target: AppAuthSwiftTests with bridging header and static linking. I believe you can do the same thing with any target, including standalone app.

In this simple example you can also notice issues like compiler warnings or unnecessary unwrapping mentioned in this ticket

**Comment by StevenEWright** [comment]
@mkujalowicz @jchnxu I suspect we will want an example that's equivalent to the Objective-C sample... but we should include something like the swift here:

https://github.com/mkujalowicz/AppAuth-iOS/blob/swift_tests/AppAuthSwiftTests/AppAuthSwiftTests.swift 

in a README / in the documentation somewhere prominent to get people going quickly.

**Comment by Laptopmini** [comment]
Bump.

I just fetched 0.7.1 and I'm getting the same warnings.

**Comment by WilliamDenniss** [comment]
Re-opening, as #79 fixed the warnings, but not the initializers that don't need to be nullable.

Added #81 to fix that, and add a Swift test.

**Comment by MichaelWTA** [comment]
Getting the same issue in `OIDExternalUserAgentIOS` now. Trying the same solution from the PRs that @WilliamDenniss linked to see if it fixes it in my project.

**Comment by jongarate** [comment]
As far as I can see, d84c0335558b4af7ab21ccd67c73d03427acef38 reverted the `nullability` again in May this year after it was merged in #81 in 2017.

What was the reason behind this? Am I missing something?

**Comment by siuying** [comment]
Saw this error lately
```
❌  BuildProductsPath/Release-iphoneos/AppAuth/AppAuth.framework/Headers/OIDExternalUserAgentIOS.h:32:1: conflicting nullability specifier on return types, 'nullable' conflicts with existing specifier 'nonnull'

```
